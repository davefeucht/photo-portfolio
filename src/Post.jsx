import React from "react";
import axios from "axios";
import PropTypes from "prop-types";

class Post extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      thumbnailUrl: "",
      fullImageUrl: "",
      showFullImage: false,
      inList: true,
      errorMsg: ""
    };
  }

  _getPostImage() {
    axios.get("http://" + this.props.site + "/wp-json/wp/v2/media/" + this.props.image + "/")
      .then(res => {
        let thumbnailImage = new Image();
        let fullImage = new Image();
        const thumbnailUrl = res.data.media_details.sizes.thumbnail.source_url;
        const fullImageUrl = res.data.media_details.sizes.full.source_url;
        this.setState({ thumbnailUrl });
        this.setState({ fullImageUrl });
        thumbnailImage.src = thumbnailUrl;
        fullImage.src = fullImageUrl;
      }, error => {
        const errorMsg = "Did not work: " + (error.response ? error.response : error);
        this.setState({ errorMsg });
      });
  } 

  _toggleDisplay(event) {
    event.preventDefault();
    this.props.clickImage(this.props.id);
    this.setState({showFullImage: !this.state.showFullImage}); 
  }

  componentWillMount() {
    this._getPostImage();
  }

  render() {
    
    let imgTag = "";
    let title = "";

    if(this.props.context == "thumbnail") {
      title = <div className="post-title-multiple"><a href="#" onClick={this._toggleDisplay.bind(this)}>{this.props.title}</a></div>;
      if(this.state.showFullImage) {
        imgTag = <div className="post-image"><img src={this.state.fullImageUrl} /></div>;
      }
      else {
        imgTag = <div className="post-image"><img src={this.state.thumbnailUrl} /></div>;
      }
    }
    else if(this.props.context == "category-image") {
      title = "";
      imgTag = <div className="post-image"><img src={this.state.thumbnailUrl} /></div>;
    }
    else {
      title = <div className="post-title-single"><a href="#" onClick={this._toggleDisplay.bind(this)}>{this.props.title}</a></div>;
      imgTag = <div className="post-image"><img src={this.state.fullImageUrl} /></div>;
    }

    return(
      <div className="post">
        {title}
        {imgTag}         
      </div>
    );

  }  

}

Post.propTypes = {
  id: PropTypes.number,
  title: PropTypes.string,
  image: PropTypes.number,
  context: PropTypes.string,
  site: PropTypes.string,
  clickImage: PropTypes.func
};

export default Post;
