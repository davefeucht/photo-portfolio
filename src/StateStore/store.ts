import { configure, makeAutoObservable } from 'mobx';
import { ImageData, Image, Store, Category, Post, Page } from '../utils/types';

configure({ enforceActions: 'observed' });

class stateStore implements Store {
    menuState = 'closed'
    applicationRoot: HTMLElement = null;
    screenInfo: {
        width: number,
        height: number
    }
    siteInfo: {
        siteName: string,
        siteUrl: string
    }
    visiblePost: {
        postId: number,
        postTitle: string,
        tags: number[],
        tagNames: string[],
        fullImageUrl: string,
        width: number,
        height: number
    };
    categoryList: Category[] = [];
    currentCategoryPosts: Post[] = [];
    currentCategoryData: Category = {
        id: 0,
        count: 0,
        description: '',
        link: '',
        name: '',
        slug: '',
        taxonomy: '',
        parent: 0
    };
    pages: Page[] = [];
    currentPageData: Page = {
        id: 0,
        link: '',
        modified: '',
        slug: '',
        type: '',
        title: {
            rendered: ''
        },
        content: {
            rendered: '',
            protected: false
        },
        author: 0,
        featured_media: 0
    };
    maxItemsPerPage = 10;

    constructor() {
        makeAutoObservable(this);
    }

    setMenuState = (state: string) => {
        this.menuState = state;
    }

    toggleMenuState = () => {
        this.menuState = this.menuState === 'closed' ? 'open' : 'closed';
    }

    setApplicationRoot = (element: HTMLElement) => {
        this.applicationRoot = element;
    }

    setSiteName = (name: string) => {
        this.siteInfo.siteName = name;
    }

    setVisiblePost = (postId: number, postTitle: string) => {
        this.visiblePost.postId = postId;
        this.visiblePost.postTitle = postTitle;
    }

    setVisiblePostImage = (fullImageUrl: string) => {
        this.visiblePost.fullImageUrl = fullImageUrl;
    }

    setVisiblePostTags = (tagNames: string[]) => {
        tagNames.forEach(tagName => {
            this.visiblePost.tagNames.push(tagName);
        });
    }

    clearVisiblePostTagNames = () => {
        this.visiblePost.tagNames.splice(0, this.visiblePost.tagNames.length);
    }

    setCategoryList = (categories: Category[]) => {
        this.categoryList.length = 0;
        categories.forEach((category, index) => {
            this.categoryList[index] = category;
        })
    }

    setCategoryPosts = (posts: Post[]) => {
        this.currentCategoryPosts.length = 0;
        posts.forEach((post, index) => {
            this.currentCategoryPosts[index] = post;
        })
    }

    setCategoryData = (categoryData: Category) => {
        Object.keys(categoryData).forEach((key: keyof Category) => {
            this.currentCategoryData[key] = categoryData[key];
        });
    }

    setThumbnailImageUrl = (imageData: ImageData) => {
        this.currentCategoryPosts[imageData.post_index].thumbnail_image = imageData.image_url;
    }

    setCurrentPost = (postData: Post) => {
        Object.keys(postData).forEach((key: keyof Image) => {
            this.visiblePost[key] = postData[key];
        })
    }

    setPages = (pages: Page[]) => {
        pages.forEach(page => {
            this.pages.push(page);
        });
    }

    setPageData = (pageData: Page) => {
        Object.assign(this.currentPageData, pageData);
    }
}

export default stateStore;
