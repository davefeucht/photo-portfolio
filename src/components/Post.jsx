/****************
* Post component displays one individual Post
****************/

import React from "react";
import axios from "axios";
import PropTypes from "prop-types";
import FullPost from "./styledComponents/FullPost.jsx";
import {connect} from "react-redux";
import {bindActionCreators} from "redux";
import {setPostUrls} from "../actions/actions.js";

class Post extends React.Component {
  static propTypes = {
    id: PropTypes.number,
    title: PropTypes.string,
    image: PropTypes.number,
    context: PropTypes.string,
    site: PropTypes.string,
    clickImage: PropTypes.func
  };

  _getPostImage() {
    const getPostImageURI = `https://${this.props.siteUrl}/wp-json/wp/v2/media/${this.props.image}`; 
    axios.get(getPostImageURI)
      .then(res => {
        let thumbnailImage = new Image();
        let fullImage = new Image();
        const thumbnailUrl = res.data.media_details.sizes.thumbnail.source_url;
        const fullImageUrl = res.data.media_details.sizes.full.source_url;
        this.props.setPostUrls({ thumbnail_image: thumbnailUrl, full_image: fullImageUrl });
        thumbnailImage.src = thumbnailUrl;
        fullImage.src = fullImageUrl;
      });
  } 

  _showAllPosts(event) {
    event.preventDefault();
    this.props.clickImage(this.props.id);
  }

  componentWillMount() {
    this._getPostImage();
  }

  render() {
    
    let divStyle = {backgroundImage: "url(" + this.props.fullImageUrl + ")", backgroundRepeat: "no-repeat", backgroundPosition: "center center", backgroundSize: "contain"};
    let classList = "post";

    return(
      <FullPost className={classList} style={divStyle} onClick={this._showAllPosts.bind(this)}>
      </FullPost>
    );
  }  
}

const mapStateToProps = state => {
  return {
    thumbnailImageUrl: state.applicationState.singlePostToShow.thumbnail_image,
    fullImageUrl: state.applicationState.singlePostToShow.full_image,
    siteUrl: state.applicationState.siteUrl,
    showAllPosts: state.visibilityFilter.showAllPosts
  };
}

const mapDispatchToProps = dispatch => {
  return {
    setPostUrls: bindActionCreators(setPostUrls, dispatch)
  }; 
}

export default connect(mapStateToProps, mapDispatchToProps)(Post);