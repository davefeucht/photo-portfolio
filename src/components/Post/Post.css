.post {
  position: absolute;
  width: 80%;
  height: 80%;
  cursor: pointer;
  grid-column: 1 / 6;
  grid-row: 1/3;
}

.post-background {
  position: absolute; 
  top: 0; 
  left: 0; 
  width: 100%;
  height: 100%; 
  background-color: rgba(0, 0, 0, 0.8); 
  z-index: 10;
}