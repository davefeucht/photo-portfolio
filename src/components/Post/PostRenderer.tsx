import PostFooter from 'components/PostFooter/PostFooter';
import PostImage from 'components/PostImage/PostImage';
import PostTitlebar from 'components/PostTitlebar/PostTitlebar';
import { reaction } from 'mobx';
import * as React from 'react';
import { useEffect, useRef, useState } from 'react';
import { setPostRect } from 'utils/PostHelper';
import { ScreenInfo, VisiblePost } from 'utils/types';

interface Props {
    onClose(): void;
    previousPost: number;
    nextPost: number;
    screenInfo: ScreenInfo;
    visiblePost: VisiblePost;
}

const image = document.createElement('img');

const PostRenderer: React.FC<Props> = ({
    onClose,
    previousPost,
    nextPost,
    screenInfo,
    visiblePost
}) => {
    const [imageHeight, setImageHeight] = useState<number>(0);
    const imageRef = useRef<HTMLImageElement>(null);

    const updateImage = () => {
        setPostRect(image, screenInfo.width, screenInfo.height, visiblePost);
    };

    image.onload = () => {
        updateImage();
        setImageHeight(imageRef.current?.clientHeight ?? 0);
    };

    useEffect(() => {
        if (visiblePost.fullImageUrl) {
            image.src = visiblePost.fullImageUrl;
        }

        if (imageRef.current) {
            imageRef.current.addEventListener('transitionend', updateImage, true);
        }

        const disposer = reaction(
            () => [screenInfo.width, screenInfo.height],
            () => updateImage()
        );

        updateImage();

        return () => {
            if (imageRef.current) {
                imageRef.current.removeEventListener('transitionend', updateImage);
            }
            disposer();
        };
    }, [visiblePost.fullImageUrl]);

    return (
        <div className="post-background" onClick={onClose}>
            <div className="post" onClick={event => event.stopPropagation()}>
                <PostTitlebar postTitle={visiblePost.postTitle} />
                <PostImage
                    imageUrl={visiblePost.fullImageUrl}
                    imageHeight={imageHeight}
                    previousPost={previousPost}
                    nextPost={nextPost}
                    ref={imageRef}
                />
                <PostFooter tagNames={visiblePost.tagNames} />
            </div>
        </div>
    );
};

export default PostRenderer;
