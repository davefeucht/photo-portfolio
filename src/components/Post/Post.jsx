/****************
* Post component displays one individual Post
****************/

import React from "react";
import axios from "axios";
import { observer } from 'mobx-react';
import './Post.css';

const Post = observer(({ props }) => {
  const _getPostImage = () => {
    const getPostImageURI = `https://${props.siteUrl}/wp-json/wp/v2/media/${this.props.image}`; 
    axios.get(getPostImageURI)
      .then(res => {
        let thumbnailImage = new Image();
        let fullImage = new Image();
        const thumbnailUrl = res.data.media_details.sizes.thumbnail.source_url;
        const fullImageUrl = res.data.media_details.sizes.full.source_url;
        props.setFullImageUrl(fullImageUrl);
        thumbnailImage.src = thumbnailUrl;
        fullImage.src = fullImageUrl;
      });
  } 

  const _showAllPosts = (event) => {
    event.preventDefault();
    props.clickImage(this.props.id);
  }
    
 _getPostImage();

  let divStyle = {backgroundImage: "url(" + props.fullImageUrl + ")", backgroundRepeat: "no-repeat", backgroundPosition: "center center", backgroundSize: "contain"};
  let classList = "post";

  return(
    <div className={classList} style={divStyle} onClick={_showAllPosts.bind(this)}>
    </div>
  );
});

Post.displayName = 'Post';

export default Post;