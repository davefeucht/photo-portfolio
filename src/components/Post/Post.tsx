/** **************
* Post component displays one individual Post in a modal
*************** */

import './Post.css';

import { reaction } from 'mobx';
import { observer } from 'mobx-react';
import * as React from 'react';
import { useEffect, useRef, useState } from 'react';
import { useNavigate, useParams } from 'react-router-dom';

import { getPost, getPostImage, getTagNames } from '../../utils/Api';
import {
    getNextPost,
    getPreviousPost,
    setPostRect
} from '../../utils/PostHelper';
import {
    Post as PostState,
    ScreenInfo,
    SiteInfo,
    VisiblePost
} from '../../utils/types';
import PostFooter from '../PostFooter/PostFooter';
import PostImage from '../PostImage/PostImage';
import PostTitlebar from '../PostTitlebar/PostTitlebar';

const image = document.createElement('img');

interface PostProps {
    screenInfo: ScreenInfo,
    siteInfo: SiteInfo,
    visiblePost: VisiblePost,
    currentCategoryPosts: PostState[],
    clearVisiblePostTagNames: () => void,
    setCurrentPost: (a: PostState, b: string[], c: string) => void
}

const Post: React.FC<PostProps> = ({
    screenInfo,
    siteInfo,
    visiblePost,
    currentCategoryPosts,
    clearVisiblePostTagNames,
    setCurrentPost
}) => {
    const [imageHeight, setImageHeight] = useState<number>(0);
    const { categoryId = '-1', postId = '-1' } = useParams();
    const navigate = useNavigate();
    const imageRef = useRef<HTMLImageElement>(null);

    const closeModalHandler = () => {
        navigate(`/category/${categoryId}`);
    };

    const stopPropagation = (e: React.MouseEvent) => {
        e.stopPropagation();
    };

    const updateImage = () => {
        setPostRect(image, screenInfo.width, screenInfo.height, visiblePost);
    };

    image.onload = () => {
        updateImage();
        setImageHeight(imageRef.current?.clientHeight ?? 0);
    };

    useEffect(() => {
        let tagNames: string[] = [];
        let url = '';
        if (postId) {
            clearVisiblePostTagNames();
            getPost(parseInt(postId), siteInfo.siteUrl)
                .then((post: PostState) => {
                    return post;
                })
                .then((newPost: PostState) => {
                    getTagNames(newPost.tags, siteInfo.siteUrl)
                        .then(tags => {
                            tagNames = tags;
                        });
                    getPostImage(newPost.featured_media, siteInfo.siteUrl)
                        .then((imageUrl: string) => {
                            url = imageUrl;
                            setCurrentPost(newPost, tagNames, url);
                        });
                });
        }
    }, [postId]);

    useEffect(() => {
        if (visiblePost.fullImageUrl) {
            image.src = visiblePost.fullImageUrl;
        }

        if (imageRef.current) {
            imageRef.current.addEventListener('transitionend', updateImage, true);
        }

        const disposer = reaction(
            () => [screenInfo.width, screenInfo.height],
            () => updateImage()
        );

        updateImage();

        return () => {
            if (imageRef.current) {
                imageRef.current.removeEventListener('transitionend', updateImage);
            }
            disposer();
        };
    }, [visiblePost.fullImageUrl]);

    return (
        <div className="post-background" onClick={closeModalHandler}>
            <div className="post" onClick={stopPropagation}>
                <PostTitlebar postTitle={visiblePost.postTitle} />
                <PostImage
                    imageUrl={visiblePost.fullImageUrl}
                    imageHeight={imageHeight}
                    previousPost={getPreviousPost(parseInt(postId), currentCategoryPosts)}
                    nextPost={getNextPost(parseInt(postId), currentCategoryPosts)}
                    ref={imageRef}
                />
                <PostFooter tagNames={visiblePost.tagNames} />
            </div>
        </div>
    );
};

Post.displayName = 'Post';

export default observer(Post);
