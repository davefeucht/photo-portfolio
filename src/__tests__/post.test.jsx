import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

import Post from '../components/Post/Post';
import StateStore from '../StateStore/store';

const post = {
    id: 887,
    featured_media: 142,
    categories: [220]
};

const tagNames = ['tag1', 'tag2', 'tag3'];

const href = 'https://fooimage';

jest.mock('../utils/Api', () => ({
    getPost: () => Promise.resolve(post),
    getTagNames: () => Promise.resolve(tagNames),
    getPostImage: () => Promise.resolve(href)
}));

test('Post displays', () => {
    const store = new StateStore();
    const component = renderer.create(
        <MemoryRouter initialEntries={['/page/150']}>
            <Post stateStore={store} />
        </MemoryRouter>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
