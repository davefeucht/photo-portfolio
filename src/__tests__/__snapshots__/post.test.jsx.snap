// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post displays 1`] = `
<div
  className="post-background"
  onClick={[Function]}
>
  <div
    className="post"
    onClick={[Function]}
  >
    <div
      className="post-titlebar"
    >
      <div
        className="title"
      />
      <div
        className="close-button"
      >
        <a
          href="/category/undefined"
          onClick={[Function]}
        >
          <div
            className="close-button-content"
          >
            <img
              height="30px"
              src="./assets/images/close-button.svg"
              width="30px"
            />
          </div>
        </a>
      </div>
    </div>
    <div
      className="post-image"
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <img
        src={null}
      />
    </div>
    <div
      className="post-footer"
    >
      <div
        className="labels"
      />
    </div>
  </div>
</div>
`;
